[#2417] Add caliper tool to the design view

SiboVG opened the pull request:
This PR fixes #2417 and adds a caliper measurement tool to the rocket design view. It also fixes a bug where the ruler in the design view would not update when switching views.

## Tool Location

The tool is accessible as a button from the RocketPanel ribbon. Clicking it, opens the Caliper Tool Dialog:

https://github.com/user-attachments/assets/239d9d80-514d-4470-8fc1-120551659b78

I initially tried to fit the widgets in the ribbon, but it cluttered the UI too much, which would especially be an issue for smaller screens. Also, in a lot of CAD software, measurement tools open up in a new dialog.

## Caliper Tool Dialog

Let's go over the different settings:
<img width="1551" height="1005" alt="image" src="https://github.com/user-attachments/assets/04c6e74c-8622-47af-9207-08d4553551a7" />

- `Caliper mode`: the caliper can be used in two modes - Vertical and Horizontal. In vertical mode, the distance is measured between 2 vertical caliper lines. In horizontal mode, it uses horizontal lines.
- `Distance`: this is the distance between the caliper lines, i.e. the measurement value you're interested in.
- `Line 1 position`: the position of caliper line 1.
  - The spinner shows the position, but can also be edited to manually enter the desired position.
  - The `Snap to geometry` button allows you to select lines and points in the design view to snap the caliper line to — more on this later.
- `Line 2 position`: same, for caliper line 2.
- `About button`: click this button to show help text on how to use the caliper:
<img width="403" height="370" alt="image" src="https://github.com/user-attachments/assets/c4c84543-ea0f-406c-a4a8-69ec182e46e6" />

(Full about text:
```
You can adjust the caliper line positions in several ways:
• Click and drag the caliper lines directly in the rocket view
• Manually change the line positions using the position spinners
• Click the "Snap to geometry" button to enter snap mode, then select an edge or point in the rocket figure
• Hold Shift and drag a caliper line to snap it to nearby geometry targets)
```

- `Minimize window`: if you click this button, the caliper tool will only show the distance field. If you double-click on the dialog background, close and reopen the dialog, or click the caliper tool button again, the dialog returns to full widget mode:
<img width="1554" height="1004" alt="image" src="https://github.com/user-attachments/assets/45f737ac-ecb4-45d3-a818-4d51322ba75b" />

- `Close`: close the dialog and exit the caliper tool.

## UI

The proposed UI in #2417 would be difficult to implement, needing swing overlays etc. This PR keeps the caliper lines confined within the design view, and displays the results in the caliper dialog. Displaying the distance inside the design view _is_ possible, but I fear that this can clash with the other design elements there. And you would probably want a way to move the text in case it does overlap, adding more complexity...

The caliper consists of two caliper lines, number 1 and 2. You can move the line by clicking and dragging anywhere on the line (see next section). To make it clearer to users that you can drag the line, a handle is drawn on top, which also includes the line number.

## Operation modes

### Design Views

The caliper only works in 2D views. The caliper settings in the side view (including top view) and the back view are independent. So, if you switch between those views, you will notice that the caliper settings change.

### Horizontal vs Vertical

As mentioned before, you can switch between vertical lines or horizontal lines.

### Click-drag (free) mode

The first mode is _free_ mode, where you can click-drag the caliper lines in the design view:

https://github.com/user-attachments/assets/fec711fa-7e87-4179-b028-4291c4e07988

### Shift-click drag snap mode

In click-drag mode, you can press the Shift key to go to _drag snap_ mode. In this mode, when you are close to a defining edge or point on the geometry (more on this later), it will snap to that feature. If not, you move in free mode. Notice how the info message below the design view changes when you drag one of the caliper lines.

https://github.com/user-attachments/assets/1f37da91-c95e-45d8-bdd7-f7ca19d28e9e

### Snap mode

If you click the `Snap to geometry` button in the caliper dialog, you can click on a defining geometry feature in the design view to snap the corresponding cursor to that feature position. These features depend on the caliper mode: horizontal or vertical. For instance, if you're in vertical mode, the vertical lines of a body tube will be selectable. For horizontal modes, the horizontal lines are. For a nose cone, the tip of the cone is selectable in both vertical and horizontal mode, whereas the base is only selectable in vertical mode. An example of a debug render which shows you all the selectable features in "A simple model rocket" in vertical caliper mode:
<img width="1554" height="1001" alt="Screenshot 2025-11-23 at 01 25 30" src="https://github.com/user-attachments/assets/ea44b79b-f494-47da-b877-d89076f51ba1" />

And horizontal mode:
<img width="1553" height="1007" alt="Screenshot 2025-11-23 at 01 25 38" src="https://github.com/user-attachments/assets/c82030bf-39aa-4111-97a0-219075d1c204" />

Under normal operation, when you hover over a geometry feature, it will render the selectable feature:

https://github.com/user-attachments/assets/305a63f7-6f72-4197-8b02-aafe53224bdf

You can also select the CG and CP as a snap target.

You can exit out of the snap mode by clicking on empty space, pressing the escape key, or by clicking the `Snap to geometry` button again. Once again, this information is displayed in the info message below the design view (except for the "Click the button again" part, I don't feel that's necessary).

### Manual caliper position

You can also enter the caliper positions manually in the spinners:

https://github.com/user-attachments/assets/66ff8310-e445-426e-b9ca-fb03cf69e959

## Caliper line out of view

It's possible for the caliper line to be out of view, e.g. when zoomed in quite far or due to bad default line positioning. In that case, a triangle indicator is rendered. If you click that indicator, the caliper line is brought back into view. Or you can manually edit the line position.

https://github.com/user-attachments/assets/974c3d8d-1e3b-4662-a98e-4ed31966008c

## Preferences

There is a design preference to by default open the caliper tool minimized:
<img width="834" height="769" alt="image" src="https://github.com/user-attachments/assets/26f9a327-ce51-472e-9d3e-27026fad3295" />

---

Enjoy :)

SiboVG committed:
[#2417] Add vertical caliper to design view

SiboVG committed:
Hide caliper in 3D view

SiboVG committed:
Add caliper line position widgets

SiboVG committed:
Rename calipers to 1 and 2

SiboVG committed:
Fix caliper handle not drawing after window resize

SiboVG committed:
Use different caliper state for back side view

SiboVG committed:
Update rulers on view change

SiboVG committed:
Refactor UI

SiboVG committed:
Improve backside caliper init positions

SiboVG committed:
Fix empty bounds check

SiboVG committed:
[WIP] Fix handle visual glitch

SiboVG committed:
Trigger repaint after resize and view switch

SiboVG committed:
Add horizontal calipers

SiboVG committed:
Refactor caliper logic into separate manager

SiboVG committed:
Fix initial view being null

SiboVG committed:
Remove import

SiboVG committed:
Add caliper line snapping to body tube

SiboVG committed:
Properly provide class

SiboVG committed:
Improve defaults

SiboVG committed:
Add caliper line out-of-view indicator

SiboVG committed:
Increase snap line width

SiboVG committed:
Exit snap mode when the rocket has changed

SiboVG committed:
Make body tube provider general for coaxial

SiboVG committed:
Use true distance to line/point

SiboVG committed:
Add transition shape provider

SiboVG committed:
Always show 4 points for circle snap

SiboVG committed:
Exit snap mode when clicking on empty target

SiboVG committed:
Listen for escape on any widget

SiboVG committed:
Move caliper widgets to dedicated panel instead of RocketPanel ribbon

SiboVG committed:
Use textfield for distance results + bigger and colored font

SiboVG committed:
Use modeless dialog
Otherwise the snap click doesn't work anymore

SiboVG committed:
Reset caliper window on button click

SiboVG committed:
Make it more visually clear which caliper pos you're editing

SiboVG committed:
Add debug mode to show all snap targets

SiboVG committed:
Fix rotation issue on tube snaps

SiboVG committed:
Exit snap mode when closing dialog

SiboVG committed:
Update snap targets on view change

SiboVG committed:
When switching from vertical to horizontal caliper, exit out of snap mode

SiboVG committed:
Don't close dialog with escape

SiboVG committed:
Add snap mode escape info

SiboVG committed:
Change color for top text

SiboVG committed:
Add button to minimize the caliper dialog

SiboVG committed:
Add fin set snap provider

SiboVG committed:
Add mass object snap target

SiboVG committed:
Add rail button provider

SiboVG committed:
Fix snap mode escape exit

SiboVG committed:
Replace snap icon

SiboVG committed:
Still show CG and CP in snap mode

SiboVG committed:
Add CG/CP snap targets

SiboVG committed:
Increase pos width

SiboVG committed:
Add snapping using shift-key

SiboVG committed:
Draw target that's used for shift-snapping

SiboVG committed:
Display help text for dragging caliper line

SiboVG committed:
Move instructions to infoMessage

SiboVG committed:
Change naming

SiboVG committed:
Add help text to caliper dialog

SiboVG committed:
Add pref to open caliper dialog minimized

SiboVG committed:
Remember caliper dialog position

SiboVG committed:
Add unit tests

SiboVG committed:
Use escape key to close the caliper dialog when not in snap mode

SiboVG committed:
Bring back the caliper line when clicking on its indicator

SiboVG committed:
Update position models when switching mode

SiboVG committed:
Remove unused imports

neilweinstock commented:
Wow, you're on quite a roll lately.  It looks like you've covered just about every possible use case here.

Just one question: when calipers are active are other functions still available (like, for instance, clicking and editing components?

SiboVG commented:
> Just one question: when calipers are active are other functions still available (like, for instance, clicking and editing components?

No, if you do that, it will exit caliper mode.

neilweinstock commented:
This would be my only reservation. But I’m going to play with it for a while to see if it matters.

neilweinstock commented:
I played with it a bit on Windows and it is very impressive.  Some comments:

1) When I enter snap to geometry mode, some yellow text appears at the top of the rocket figure which is unreadable in light mode:
<img width="1353" height="1007" alt="image" src="https://github.com/user-attachments/assets/53f94348-45c0-4e21-a4e4-8ada56923dd6" />

2) I am unable to click and drag the calipers.  I can move them with the spinners and I can click on a geometry point to snap to geometry, but I can't just drag them around.  When I mouse over one it will thicken to highlight, but that's about it.

3) Caliper mode only becomes modal when I click snap to geometry.  I didn't realize how it worked: it allows you a single geometry click then it deactivates.  On the one hand it is good that I can still edit the design normally in caliper mode; on the other I feel like some will want to stay continuously in snap mode, which they can't do right now.  I'm still thinking about this, no recommendations yet.

4) The Caliper tool window has a mixture of controls and indicators.  I think it would be better to separate them.  For instance, I'd like the distance between them to be shown in the rocket figure display.  As for the line positions.... I'd probably prefer them in the rocket figure as well, although that then makes it harder to know what to do with the spinners and snap buttons.  I will think about this some more.  Yes I know you can minimize the window, but that doesn't solve it for me.

5) I think the "1" and "2" in the calipers are too large, I'd make the font somewhat smaller.  They seem out of proportion compared to the rest of the display.

6) I also want to think more about the "Show CG/CP" and "Caliper Tool" buttons.  They both have a label above them that is pretty much redundant.  Wondering if we could stack these two elements and get rid of the labels, since the buttons are self-explanatory.  Will fiddle with some mockups.

Fantastic work overall.

jimmiedave commented:
This looks like a thing of beauty!

I'm going to try to play with it in the next day or two.

jimmiedave commented:
Wow, that is really great. I'm playing with it on macOS right now.

I like the way it works in general. Echoing @neilweinstock , fantastic work!

To Neil's comment about "staying in Snap mode" - hold [SHIFT] to do so. Unless you mean "stay in Snap with no modifier keys", which...nope, not currently.

Very few quibbles from me:

1. bug-ish behavior: in Top View, the highlight of circular items in Snap mode differs from the side view. Not the biggest deal in the world, but should be observed, thought about.
2. Not apparent how to un-minimize the minimized caliper window. Could use a button or window icon or something. macOS used to use the green button for this, but Apple F-ed it up pretty bad.
3. Speaking of the green button, the layout looks extremely hinky when you hit Maximize. That little window's guts spilled over a 27in 4K monitor is...troubling. Should maybe eliminate that possibility (no green button) or cram it all top left or something.

A UI thought: should the buttons and widgets layout match the caliper lines' organization on the screen? f.e. defaults to vertical lines measuring left-right, but the widgets for these lines are top-bottom, not left-right columns. No idea whether it'd be weird to then have that layout swap when the lines are swapped to horizontal lines (where over-under layout reflects the screen layout). Is this clear, or a cloudy mess of a suggestion?

For me, if it were built as-is today, I'd be thrilled. @neilweinstock 's comments are well-thought-out (though the numbers look about right-sized to me on Mac).

VERY happy to have this feature see the light of day. Props for excellent work, @SiboVG !

JoePfeiffer commented:
I'll echo @neilweinstock and @jimmiedave -- this is amazing. I do have some suggestions:

1. I'd also like to have a way to go into snap mode and just leave it there. In my own imagined use case, I'd want to click the "snap mode" button and have it keep snapping to the nearest geometry element until I leave snap mode.
2. I agree that being able to "drag" the lines would be very welcome. If not in snap mode they move to exactly where the user drags them; if in snap mode the user drags them to the desired feature and they snap into place when the mouse button is released.
3. The faint yellow outline around the feature to be snapped is extremely welcome, and would also be helpful in selecting elements when in editing mode.
4. Something very minor: while having a vertical or horizontal caliper mode is logically consistent, it seems more natural to me to call it "measurement mode" and have the meaning reversed (ie horizontal measurement mode has vertical calipers and makes horizontal measurements).

jimmiedave commented:
Agree with @JoePfeiffer 's #1 on Snap mode - and I'm not sure the snap buttons then have value since you can drag to snap in snap mode. Could simplify the UI just turning Snap on or off.

I can drag the lines just fine - unsure what the trouble would be, though I'm on macOS - maybe a platform difference? Or maybe I'm not getting what's being described.

A behavior note: with the horizontal-line calipers set to, f.e. the extents of a body tube in side or top view, switching to the back view doesn't retain the caliper location. I can see that this might be problematic, given the ability to rotate these views, though you could, when switching, retain the rotation setting, apply the same rotation to the back view and retain and apply the caliper setting to the back view (for horizontal-line calipers - vertical-line calipers wouldn't provide anything sensible in this transition). Maybe too complicated to solve for this - it works fine as is now.

In writing the previous paragraph, it became clear that we probably need better names for these, since horizontal calipers measure vertical distance and vice versa. Should probably relate them to the model's context: "Longitudinal" and "transverse" sound pretty windy and unclear. Maybe "length caliper" and "width caliper"? Better options? Both would be "width calipers" in back view.

"Snap to Geometry" button doesn't want to snap to the curve of a nose cone. It's not clear what that would mean (where the snap would occur), but it took a few moments of thinking about it before I realized that was why the nose cone wouldn't highlight for snap. Maybe one way to deal with this is to snap to the greatest extent, or to the centerline of a nose cone, whichever is nearer to the click.

The "1" and "2" numbers are big-ish (compared to the rulers) for me, but not huge. Not sure why @neilweinstock 's experience is different from mine. Using this on a 14in MacBook Pro, retina display.

<img width="325" height="298" alt="image" src="https://github.com/user-attachments/assets/2f1b5193-b74a-4706-bd84-5ed244ab6645" />

SiboVG commented:
>     1. When I enter snap to geometry mode, some yellow text appears at the top of the rocket figure which is unreadable in light mode:

Yeah, I only work in dark mode; I should switch to light mode more often during testing 🙃 Or maybe this is part of my plan to convert everyone to dark mode hehe...

>     2. I am unable to click and drag the calipers.  I can move them with the spinners and I can click on a geometry point to snap to geometry, but I can't just drag them around.  When I mouse over one it will thicken to highlight, but that's about it.

That's weird, I can click-drag the lines. I'll investigate.

>     4. The Caliper tool window has a mixture of controls and indicators.  I think it would be better to separate them.  For instance, I'd like the distance between them to be shown in the rocket figure display.  As for the line positions.... I'd probably prefer them in the rocket figure as well, although that then makes it harder to know what to do with the spinners and snap buttons.  I will think about this some more.  Yes I know you can minimize the window, but that doesn't solve it for me.

Hm... I'm not sure. I believe having it in the rocket display would complicate things... At least I don't know how you can cleanly implement it.

>     5. I think the "1" and "2" in the calipers are too large, I'd make the font somewhat smaller.  They seem out of proportion compared to the rest of the display.

Yes, the handles and text look normal on my screen, but it should be adjusted to work on different resolutions as well.

>     6. I also want to think more about the "Show CG/CP" and "Caliper Tool" buttons.  They both have a label above them that is pretty much redundant.  Wondering if we could stack these two elements and get rid of the labels, since the buttons are self-explanatory.  Will fiddle with some mockups.

Yes, was thinking the same thing while working on this issue. I'm okay with removing the labels.

SiboVG commented:
A note though: it'll take some time before I continue work on this PR. It's been painful to get it out, so I wanna take a short break from it, and I have other stuff on my mind at work. In the meantime, you're welcome to squabble about the final details 😉

SiboVG committed:
Make caliper handle sizes and fonts DPI-aware

- Convert hardcoded pixel sizes to DPI-scaled values based on GUIUtil.getDPI()
- Base sizes defined at 96 DPI and scaled based on actual DPI
- Reduced base font sizes (30pt -> 20pt, 24pt -> 16pt) for better appearance
- Applied to handle dimensions, line widths, fonts, and arrow indicators
- Fixes issue where handles and text appeared too large on high-DPI Windows systems

SiboVG committed:
Use screen-space hit testing for caliper handle dragging

- Replace model-space tolerance with screen-space handle bounds checking
- Add getHandleBounds() methods to CaliperLine and HorizontalCaliperLine
- Add getModelToScreenTransform() helper method in CaliperManager
- Check if click is within handle bounds in screen coordinates
- Fixes dragging not working on Windows due to unreliable model-space tolerance

jimmiedave commented:
When I try to run the jar from build 2860 on Ubuntu, I get `Attempting to set up OSX properties on non-MAC_OS`, and no startup.

jimmiedave commented:
On Windows 11 ARM on VM, Apple ARM CPU, Can't drag the calipers, numbers look bigger than they ought. This also with build 2860.

SiboVG committed:
Fix caliper dragging on other platforms

SiboVG committed:
Improve caliper highlight color for light mode

neilweinstock commented:
Dragging now works under Windows, and the snap text is readable (although still a bit low-contrast) in light mode.

"1" and "2" labels in caliper handles are a bit better.

Will be back with more detailed suggestions after thinking it through.

jimmiedave commented:
Trying to run `build_2879` from the .jar, on Ubuntu Linux. Still getting:

`12:30:53.053 [main] WARN info.openrocket.swing.startup.OSXSetup -- Attempting to set up OSX properties on non-MAC_OS`

OR never starts.

Is there a command-line switch or something I can use to heal this?

JoePfeiffer commented:
It's starting fine for me, running Debian on x86_64.

Dragging is now working for me.

The "Select the edge or point..." text is now legible (in light mode), but still a bit hard to see.  Is it the same color as the component highlight? It looks a bit lighter, though that could just an illusion because it's a smaller area.

It would be nice if the distance and line position values would update as I hover across the figure.

I still think using "measurement mode" rather than "caliper mode" would be clearer.

Coming along very nicely!

jimmiedave commented:
> It's starting fine for me, running Debian on x86_64.

Then this seems likely a Linux on ARM thing. I'll try on my AMD hardware Linux machine when I get to the house.

jimmiedave commented:
Yeah, nope. This on Ubuntu 24.04.3 LTS on AMD x86_64 hardware linux box:

<img width="1190" height="43" alt="image" src="https://github.com/user-attachments/assets/0476f388-a591-4dd5-a9fa-5ee3b5b15630" />

neilweinstock commented:
While we separately debate/negotiate the UI for this thing, I have a request for a small tweak that I hope is non-controversial: right now, if both calipers are off-screen in the same direction, the off-screen indicators appear on top of each other.  It would be good if in that case they are displayed one above each other so you can see and/or click them separately.

My preferred approach would be to always display the cursor 1 indicator slightly above center, and the cursor 2 indicator slightly below, so they're always in a predictable place and inherently cannot overlap each other.

JoePfeiffer commented:
I've been using this to get measurements for rockets I'm reviewing for IREC, and it's great. I think @neilweinstock 's suggestions would also add a lot to it.

One thing, the "snap to component" function doesn't seem to work with freeform fins.

SiboVG committed:
Merge branch 'unstable' into issue-2417

# Conflicts:
#	swing/src/main/java/info/openrocket/swing/gui/scalefigure/RocketPanel.java
#	swing/src/main/java/info/openrocket/swing/gui/util/Icons.java
#	swing/src/main/java/module-info.java

SiboVG committed:
Use ruler SVG icon

SiboVG committed:
Merge branch 'unstable' into issue-2417

# Conflicts:
#	swing/src/main/java/info/openrocket/swing/gui/scalefigure/RocketPanel.java
#	swing/src/main/java/info/openrocket/swing/gui/util/Icons.java

SiboVG committed:
Fix small merge conflict

SiboVG committed:
Change caliper handle to diamond shape

SiboVG committed:
[WIP] Rework UI

SiboVG committed:
Fix checkbox snap behavior

SiboVG committed:
Add snap icon

SiboVG committed:
Use togglebutton with icon for calipers

SiboVG committed:
Use togglebutton with icon for snap

SiboVG committed:
Grey out icons when selected

SiboVG committed:
Fix translation

SiboVG committed:
Remove unused UI code

SiboVG committed:
Add popup menu for caliper lines

SiboVG committed:
Improve clarity of distance value reading

SiboVG committed:
Improve caliper colors (for color-blind people)

SiboVG committed:
Use filled magnet icon

SiboVG committed:
Improve colors

SiboVG committed:
Decrease caliper handle size

SiboVG committed:
Make it clearer that handle icons can be clicked

SiboVG committed:
Improve hover effect of caliper lines

SiboVG committed:
Add click hint when caliper indicator out-of-bounds

SiboVG committed:
Fix click priority of out-of-bounds caliper

SiboVG committed:
Use editable spinner

SiboVG committed:
Add hint for exiting click-snap mode

SiboVG committed:
Add top relative rendering for caliper
To make sure calipers are always drawn on top

SiboVG committed:
Reduce gap before stage selector separator

SiboVG committed:
Use "Caliper" instead of "Calipers"

SiboVG committed:
Add SVG icon of snap click

SiboVG committed:
Add snap highlight to top relative

SiboVG commented:
New proposal:

https://github.com/user-attachments/assets/2d92b9c8-b7ed-44c2-ae0b-9db70a034156

neilweinstock commented:
This is excellent work, and definitely good enough for first beta, whenever that is.  I have only a few very small comments.

1. Very small point: for the horizontal/vertical select buttons, I think I still prefer representations of the calipers rather than the arrows.  I think it's more immediately obvious what each means.  E.g.:
<img width="426" height="69" alt="image" src="https://github.com/user-attachments/assets/44e082dc-a274-4510-a5f6-ef39951bc43c" />

2. As I commented previously (on Discord maybe?) when both calipers are off-screen, the markers are rendered on top of each other like this:
<img width="260" height="304" alt="image" src="https://github.com/user-attachments/assets/f6805190-9cd1-4fc7-97e6-4f8288dd62d2" />

The two markers should be offset around the center point so they won't overlap, and the word "click" placed dead center, like this:
<img width="519" height="608" alt="image" src="https://github.com/user-attachments/assets/64105280-2c50-45c9-94de-c69e83e709e3" />

Note that the word 'click' is cropped so it just says "lick" on my display, that should be fixed either way.

3.  I suspect it would be useful to see the numerical value of the cursor while you drag it (or just click on it or something).  I'm not sure the best way to do it, and it is not necessary for first beta.  We could just wait for feedback on this one.

Very, very, very nice work.  Users are going to love this.

SiboVG committed:
Use dual arrows for caliper mode switch

SiboVG committed:
Don't overlap calipers is out-of-view

SiboVG committed:
Ensure calipers aren't brought back on the same location

SiboVG committed:
Display caliper line position while dragging

SiboVG commented:
> This is excellent work, and definitely good enough for first beta, whenever that is. I have only a few very small comments.
> 
>     1. Very small point: for the horizontal/vertical select buttons, I think I still prefer representations of the calipers rather than the arrows.  I think it's more immediately obvious what each means.  E.g.:
> 
> 
> <img alt="image" width="426" height="69" src="https://private-user-images.githubusercontent.com/243872/567484474-44e082dc-a274-4510-a5f6-ef39951bc43c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ1NTM1MTgsIm5iZiI6MTc3NDU1MzIxOCwicGF0aCI6Ii8yNDM4NzIvNTY3NDg0NDc0LTQ0ZTA4MmRjLWEyNzQtNDUxMC1hNWY2LWVmMzk5NTFiYzQzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyNlQxOTI2NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNTJkMjAzOGI4OTVhYjhmOWZmZWUwM2VhMWE2NWEwOWMzY2EwOWU1MzkwMjVhOTFlMDc4YmIwNTdlZGE5M2VmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.anRZKQE-cFEXlqJvFaqLeG9ni20BycB76jRwqDzztks">
> 
>     2. As I commented previously (on Discord maybe?) when both calipers are off-screen, the markers are rendered on top of each other like this:
> 
> 
> <img alt="image" width="260" height="304" src="https://private-user-images.githubusercontent.com/243872/567482473-f6805190-9cd1-4fc7-97e6-4f8288dd62d2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ1NTM1MTgsIm5iZiI6MTc3NDU1MzIxOCwicGF0aCI6Ii8yNDM4NzIvNTY3NDgyNDczLWY2ODA1MTkwLTljZDEtNGZjNy05N2U2LTRmODI4OGRkNjJkMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyNlQxOTI2NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YWMwNTU1OWM5MWU1NGMyZmY2OWIxMmVhMDM0YTllZDdkNWZhM2NiMGQwNWEzOTAzODU0ZjdkODM4NTBkZjQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.i5ZDopY4HPr5vNTS-ELZ4qQIbV-LFcLmxrpgPc1miDQ">
> 
> The two markers should be offset around the center point so they won't overlap, and the word "click" placed dead center, like this: <img alt="image" width="519" height="608" src="https://private-user-images.githubusercontent.com/243872/567483647-64105280-2c50-45c9-94de-c69e83e709e3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ1NTM1MTgsIm5iZiI6MTc3NDU1MzIxOCwicGF0aCI6Ii8yNDM4NzIvNTY3NDgzNjQ3LTY0MTA1MjgwLTJjNTAtNDVjOS05NGRlLWM2OWU4M2U3MDllMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyNlQxOTI2NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOGE5YzhiZDYzOTBjMTQwZGVmNzk5NGJkYTFjYWYwZWRiMGUwZDBlN2M5ZjEwNDUxYzExOGI3Mzk4ODg0NGY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.taPJTP1Nq_qVaCCD7Hg2O7NyPeQx_h_fLUhrEJQ6A1s">
> 
> Note that the word 'click' is cropped so it just says "lick" on my display, that should be fixed either way.
> 
>     3. I suspect it would be useful to see the numerical value of the cursor while you drag it (or just click on it or something).  I'm not sure the best way to do it, and it is not necessary for first beta.  We could just wait for feedback on this one.
> 
> 
> Very, very, very nice work. Users are going to love this.

New behavior:
<img width="1555" height="1005" alt="image" src="https://github.com/user-attachments/assets/bf40a9c4-02e8-49b5-9bea-67e689c2dbbc" />

I didn't want a single "Click" because then it was less clear that you could select only one of the two calipers, that they are independent.

Position display:


https://github.com/user-attachments/assets/831cda2e-0902-4dd8-88ec-ae1864005e6b

SiboVG committed:
Merge branch 'unstable' into issue-2417

# Conflicts:
#	swing/src/main/java/info/openrocket/swing/gui/scalefigure/RocketPanel.java

SiboVG committed:
Merge branch 'unstable' into issue-2417

JoePfeiffer commented:
This is absolutely outstanding, wow. So far as I'm concerned we should merge it.

The default "snap" mode works exactly as I'd like.

SiboVG commented:
> This is absolutely outstanding, wow. So far as I'm concerned we should merge it.
> 
> The default "snap" mode works exactly as I'd like.

Thanks 🙂 . This PR has certainly been a journey.

SiboVG committed:
Reduce spacing between units and distance widgets

neilweinstock commented:
> Thanks 🙂 . This PR has certainly been a journey.

Thank you for being a willing participant on the journey. 😄  The end result is fantastic.
